ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/util/P2PDumpManager.java

```diff
@@ -0,0 +1,108 @@
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.statetransition.util;
+
+import java.io.IOException;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+
+public class P2PDumpManager {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private static final String GOSSIP_DECODING_ERROR_DIR = "gossip_decoding_error_messages";
+  private static final String GOSSIP_REJECTED_DIR = "rejected_gossip_messages";
+  private static final String INVALID_BLOCK_DIR = "invalid_blocks";
+
+  private final Path directory;
+
+  public P2PDumpManager(final Path directory) {
+    this.directory = directory;
+    if (this.directory.resolve(GOSSIP_DECODING_ERROR_DIR).toFile().mkdir()) {
+      LOG.info(
+          String.format(
+              "%s directory has been created to save gossip messages with decoding errors.",
+              GOSSIP_DECODING_ERROR_DIR));
+    }
+    if (this.directory.resolve(GOSSIP_REJECTED_DIR).toFile().mkdir()) {
+      LOG.info(
+          String.format(
+              "%s directory has been created to save rejected gossip messages.",
+              GOSSIP_REJECTED_DIR));
+    }
+    if (this.directory.resolve(INVALID_BLOCK_DIR).toFile().mkdir()) {
+      LOG.info(
+          String.format(
+              "%s directory has been created to save invalid blocks.", INVALID_BLOCK_DIR));
+    }
+  }
+
+  public String saveGossipMessageDecodingError(
+      final String topic, final String arrivalTimestamp, final Bytes originalMessage) {
+    final String fileName = String.format("%s_%s.ssz", arrivalTimestamp, topic);
+    final String identifiers = String.format("Topic: %s", topic);
+    return saveBytesToFile(
+        "gossip message with decoding error",
+        GOSSIP_DECODING_ERROR_DIR,
+        fileName,
+        identifiers,
+        originalMessage);
+  }
+
+  public String saveGossipRejectedMessageToFile(
+      final String topic, final String arrivalTimestamp, final Bytes decodedMessage) {
+    final String fileName = String.format("%s_%s.ssz", arrivalTimestamp, topic);
+    final String identifiers = String.format("Topic: %s", topic);
+    return saveBytesToFile(
+        "rejected gossip message", GOSSIP_REJECTED_DIR, fileName, identifiers, decodedMessage);
+  }
+
+  public String saveInvalidBlockToFile(
```

Comment:
When I said the Optional<String> didn't make sense, I meant that this method should not care about returning anything (fire and forget). Because nothing that happens in this method should influence other business logic, this is a cross-cutting concern.

The same applies to the other methods.